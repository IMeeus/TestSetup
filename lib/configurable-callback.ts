import { DeepPartial } from "./deep-partial";

export class ConfigurableCallback<TConfig, TResult> {
  private defaultConfig: TConfig;
  private callback: (config: TConfig) => TResult;

  constructor(defaultConfig: TConfig, callback: (config: TConfig) => TResult) {
    this.defaultConfig = defaultConfig;
    this.callback = callback;
  }

  run(override?: DeepPartial<TConfig> | ((config: TConfig) => void)): TResult {
    let config: TConfig;

    if (typeof override !== "function") {
      config = { ...this.defaultConfig, ...override };
    } else {
      config = { ...this.defaultConfig };
      override(config);
    }

    return this.callback(config);
  }
}
