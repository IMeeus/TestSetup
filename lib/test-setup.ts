import { DeepPartial } from "./deep-partial";

export class TestSetup<TConfig, TResult> {
  private defaultConfig: TConfig;
  private setupFunc: (config: TConfig) => TResult;

  constructor(defaultConfig: TConfig, setupFunc: (config: TConfig) => TResult) {
    this.defaultConfig = defaultConfig;
    this.setupFunc = setupFunc;
  }

  run(override?: DeepPartial<TConfig> | ((config: TConfig) => void)): TResult {
    let config = { ...this.defaultConfig };

    if (typeof override === "function") {
      override(this.defaultConfig);
    } else {
      config = { ...config, ...override };
    }

    return this.setupFunc(config);
  }
}
