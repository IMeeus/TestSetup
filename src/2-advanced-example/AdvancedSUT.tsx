import { useFetchPersonInfo } from "./useApi";

export interface ISUTProps {
  title: string;
}

export function SUT({ title }: ISUTProps) {
  const { isLoading, data } = useFetchPersonInfo();

  if (isLoading) {
    return <div>Loader</div>;
  }

  return (
    <>
      <h1>{title}</h1>

      <ul>
        <li>{data.firstName}</li>
        <li>{data.lastName}</li>
      </ul>
    </>
  );
}
