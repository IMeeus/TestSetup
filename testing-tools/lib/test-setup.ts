import { DeepPartial } from "./deep-partial";

export class TestSetup<TConfig, TResult> {
  private defaultConfig: TConfig;
  private setupFunc: (config: TConfig) => TResult;

  constructor(defaultConfig: TConfig, setupFunc: (config: TConfig) => TResult) {
    this.defaultConfig = defaultConfig;
    this.setupFunc = setupFunc;
  }

  run(overrides: DeepPartial<TConfig> = {}): TResult {
    return this.setupFunc({ ...this.defaultConfig, ...overrides });
  }

  run2(overrideFunc: (config: TConfig) => void = () => {}): TResult {
    const config = { ...this.defaultConfig };
    overrideFunc(config);

    return this.setupFunc(config);
  }
}
