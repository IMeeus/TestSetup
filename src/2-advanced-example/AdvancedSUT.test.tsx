import { render } from "@testing-library/react";
import { SUT, ISUTProps } from "./AdvancedSUT";
import { TestSetup } from "../../lib/test-setup";
import { ApiResponse, PersonInfo } from "./useFetchPersonInfo";
import * as useFetchPersonInfo from "./useFetchPersonInfo";

type SetupConfig = {
  props: ISUTProps;
  stubs: {
    useFetchPersonInfo: ApiResponse<PersonInfo>;
  };
};

const defaultConfig: SetupConfig = {
  props: {
    title: "someTitle",
  },
  stubs: {
    useFetchPersonInfo: {
      isOk: true,
      isLoading: false,
      data: {
        firstName: "Rubber",
        lastName: "Duck",
      },
    },
  },
};

const setup = new TestSetup(defaultConfig, (config) => {
  jest
    .spyOn(useFetchPersonInfo, "useFetchPersonInfo")
    .mockReturnValue(config.stubs.useFetchPersonInfo);

  return { ...render(<SUT {...config.props} />) };
});

test("renders correctly by default", () => {
  const { container } = setup.run();

  expect(container).toMatchSnapshot();
});

test("renders correctly when api is loading", () => {
  const { container } = setup.run({
    stubs: {
      useFetchPersonInfo: {
        isLoading: true,
      },
    },
  });

  expect(container).toMatchSnapshot();
});
