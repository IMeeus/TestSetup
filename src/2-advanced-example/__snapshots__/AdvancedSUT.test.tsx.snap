// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly by default 1`] = `
<div>
  <h1>
    someTitle
  </h1>
  <ul>
    <li>
      Rubber
    </li>
    <li>
      Duck
    </li>
  </ul>
</div>
`;

exports[`renders correctly when api is loading 1`] = `
<div>
  <div>
    Loader
  </div>
</div>
`;
